util/sock_util.c

```diff
@@ -276,3 +278,70 @@ sock_addr_prettyprint(const struct sock_addr * sa)
 		return (strdup("Unknown address"));
 	}
 }
+
+/**
+ * sock_addr_ensure_port(addr):
+ * Allocate a new string to serve as the address for sock_resolve().
+ * If ${addr} contains a port number or is the address of a Unix domain
+ * socket, duplicate that string; if not, add a port number of ":0".
+ */
+char *
+sock_addr_ensure_port(const char * addr)
+{
+	char * bind_addr;
+	char * cl;
+	char * cr;
+
+	/* Sanity check. */
+	assert(addr != NULL);
+
+	/* Where are the left-most and right-most colons in $addr? */
+	cl = strchr(addr, ':');
+	cr = strrchr(addr, ':');
+
+	/* Figure out what type of address $addr is. */
+	if (addr[0] == '/') {
+		/* It's a Unix domain socket and doesn't need a port number. */
+		bind_addr = strdup(addr);
+	} else if (addr[0] != '[') {
+		/* It's a hostname... */
+		if (cl == NULL) {
+			/* ... without any port number, so we add ":0". */
+			if (asprintf(&bind_addr, "%s:0", addr) == -1) {
+				warnp("asprintf");
+				goto err0;
+			}
+		} else {
+			/* ... which already has a port number. */
+			bind_addr = strdup(addr);
+		}
+	} else {
+		/* addr[0] == '[', so it's an address of type... */
```

Comment:
This seems like way too much work.  There's no need to distinguish between IPv4 and IPv6; we can simply check "is there a rightmost `:`, and does it have a `]` immediately before it?"

Reply:
Hmm, could I get a third opinion?  A month ago you didn't want to look for `]` and suggested `strspn()` instead:
https://github.com/Tarsnap/spiped/pull/358#discussion_r889726814

Reply:
Updated, although that ended up being a bigger change.
- added a sanity check for not going out of memory if given ":"
- added a test for "should fail" strings, namely ":" and ":1".

Reply:
PS: there's no real point to checking ":1", so I've removed it.